core/network/impl/stream_engine.hpp

```diff
@@ -143,91 +159,58 @@ namespace kagome::network {
       return add(std::move(stream), protocol, Direction::OUTGOING);
     }
 
-    outcome::result<void> add(std::shared_ptr<Stream> stream,
-                              const std::shared_ptr<ProtocolBase> &protocol) {
+    outcome::result<void> addBidirectional(
+        std::shared_ptr<Stream> stream,
+        const std::shared_ptr<ProtocolBase> &protocol) {
       return add(std::move(stream), protocol, Direction::BIDIRECTIONAL);
     }
 
-    void del(const PeerId &peer_id,
-             const std::shared_ptr<ProtocolBase> &protocol) {
-      std::unique_lock cs(streams_cs_);
-      auto peer_it = streams_.find(peer_id);
-      if (peer_it != streams_.end()) {
-        auto &protocols = peer_it->second;
-        auto protocol_it = protocols.find(protocol->protocol());
-        if (protocol_it != protocols.end()) {
-          protocols.erase(protocol_it);
-          if (protocols.empty()) {
-            streams_.erase(peer_it);
-          }
-        }
+    void reserveStreams(const PeerId &peer_id,
+                        const std::shared_ptr<ProtocolBase> &protocol) {
+      BOOST_ASSERT(protocol != nullptr);
+      auto const reserved = streams_.exclusiveAccess([&](auto &streams) {
+        return streams[peer_id]
+            .emplace(protocol->protocol(), ProtocolDescr{protocol, {}, {}, {}})
```

Comment:
`ProtocolDescr` initialization via `{}, {}, {}` does not look very nice.
An improvement could be adding custom constructors for the same behavior and cleaner look.